File: Terminal/Include/Pascal/BearLibTerminal.pas
{*
* BearLibTerminal
* Copyright (C) 2013-2017 Cfyz, Apromix
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
* of the Software, and to permit persons to whom the Software is furnished to do
* so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
* FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
* COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
* IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
* CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*}

{$H+}

unit BearLibTerminal;

interface

uses
  Types;

const
  // Keyboard scancodes
  TK_A               = $04;
  TK_B               = $05;
  TK_C               = $06;
  TK_D               = $07;
  TK_E               = $08;
  TK_F               = $09;
  TK_G               = $0A;
  TK_H               = $0B;
  TK_I               = $0C;
  TK_J               = $0D;
  TK_K               = $0E;
  TK_L               = $0F;
  TK_M               = $10;
  TK_N               = $11;
  TK_O               = $12;
  TK_P               = $13;
  TK_Q               = $14;
  TK_R               = $15;
  TK_S               = $16;
  TK_T               = $17;
  TK_U               = $18;
  TK_V               = $19;
  TK_W               = $1A;
  TK_X               = $1B;
  TK_Y               = $1C;
  TK_Z               = $1D;
  TK_1               = $1E;
  TK_2               = $1F;
  TK_3               = $20;
  TK_4               = $21;
  TK_5               = $22;
  TK_6               = $23;
  TK_7               = $24;
  TK_8               = $25;
  TK_9               = $26;
  TK_0               = $27;
  TK_RETURN          = $28;
  TK_ENTER           = $28;
  TK_ESCAPE          = $29;
  TK_BACKSPACE       = $2A;
  TK_TAB             = $2B;
  TK_SPACE           = $2C;
  TK_MINUS           = $2D;
  TK_EQUALS          = $2E;
  TK_LBRACKET        = $2F;
  TK_RBRACKET        = $30;
  TK_BACKSLASH       = $31;
  TK_SEMICOLON       = $33;
  TK_APOSTROPHE      = $34;
  TK_GRAVE           = $35;
  TK_COMMA           = $36;
  TK_PERIOD          = $37;
  TK_SLASH           = $38;
  TK_F1              = $3A;
  TK_F2              = $3B;
  TK_F3              = $3C;
  TK_F4              = $3D;
  TK_F5              = $3E;
  TK_F6              = $3F;
  TK_F7              = $40;
  TK_F8              = $41;
  TK_F9              = $42;
  TK_F10             = $43;
  TK_F11             = $44;
  TK_F12             = $45;
  TK_PAUSE           = $48;
  TK_INSERT          = $49;
  TK_HOME            = $4A;
  TK_PAGEUP          = $4B;
  TK_DELETE          = $4C;
  TK_END             = $4D;
  TK_PAGEDOWN        = $4E;
  TK_RIGHT           = $4F;
  TK_LEFT            = $50;
  TK_DOWN            = $51;
  TK_UP              = $52;
  TK_KP_DIVIDE       = $54;
  TK_KP_MULTIPLY     = $55;
  TK_KP_MINUS        = $56;
  TK_KP_PLUS         = $57;
  TK_KP_ENTER        = $58;
  TK_KP_1            = $59;
  TK_KP_2            = $5A;
  TK_KP_3            = $5B;
  TK_KP_4            = $5C;
  TK_KP_5            = $5D;
  TK_KP_6            = $5E;
  TK_KP_7            = $5F;
  TK_KP_8            = $60;
  TK_KP_9            = $61;
  TK_KP_0            = $62;
  TK_KP_PERIOD       = $63;
  TK_SHIFT           = $70;
  TK_CONTROL         = $71;
  TK_ALT             = $72;

  // Mouse events/states
  TK_MOUSE_LEFT       = $80; // Buttons
  TK_MOUSE_RIGHT      = $81;
  TK_MOUSE_MIDDLE     = $82;
  TK_MOUSE_X1         = $83;
  TK_MOUSE_X2         = $84;
  TK_MOUSE_MOVE       = $85; // Movement event
  TK_MOUSE_SCROLL     = $86; // Mouse scroll event
  TK_MOUSE_X          = $87; // Cusor position in cells
  TK_MOUSE_Y          = $88;
  TK_MOUSE_PIXEL_X    = $89; // Cursor position in pixels
  TK_MOUSE_PIXEL_Y    = $8A;
  TK_MOUSE_WHEEL      = $8B; // Scroll direction and amount
  TK_MOUSE_CLICKS     = $8C; // Number of consecutive clicks

  // If key was released instead of pressed, it's code will be OR'ed with TK_KEY_RELEASED
  TK_KEY_RELEASED     = $100;

  // Virtual key-codes for internal terminal states/variables.
  // These can be accessed via terminal_state function.
  TK_WIDTH            = $C0; // Terminal window size in cells
  TK_HEIGHT           = $C1;
  TK_CELL_WIDTH       = $C2; // Character cell size in pixels
  TK_CELL_HEIGHT      = $C3;
  TK_COLOR            = $C4; // Current foregroung color
  TK_BKCOLOR          = $C5; // Current background color
  TK_LAYER            = $C6; // Current layer
  TK_COMPOSITION      = $C7; // Current composition state
  TK_CHAR             = $C8; // Translated ANSI code of last produced character
  TK_WCHAR            = $C9; // Unicode codepoint of last produced character
  TK_EVENT            = $CA; // Last dequeued event
  TK_FULLSCREEN       = $CB; // Fullscreen state

  //Other events
  TK_CLOSE            = $E0;
  TK_RESIZED          = $E1;

  // Generic mode enum.
  // Right now it is used for composition option only.
  TK_OFF              =   0;
  TK_ON               =   1;

  // Input result codes for terminal_read function.
  TK_INPUT_NONE       =   0;
  TK_INPUT_CANCELLED  =  -1;
  
  // Text alignment.
  TK_ALIGN_DEFAULT    =   0;
  TK_ALIGN_LEFT       =   1;
  TK_ALIGN_RIGHT      =   2;
  TK_ALIGN_CENTER     =   3;
  TK_ALIGN_TOP        =   4;
  TK_ALIGN_BOTTOM     =   8;
  TK_ALIGN_MIDDLE     =  12;

// ----------------------------------------------------------------------------
// Module interface
// ----------------------------------------------------------------------------

{$IFNDEF FPC}
type
  Int32 = Integer;
  PInt32 = ^Integer;
  UInt32 = Cardinal;
  PUInt32 = ^Cardinal;
{$ENDIF}

type
  TSize = Types.TSize;

// Open
function terminal_open(): LongBool;
  cdecl; external 'BearLibTerminal' name 'terminal_open';

// Close
procedure terminal_close();
  cdecl; external 'BearLibTerminal' name 'terminal_close';

// Set
function terminal_set(const Options: AnsiString): LongBool; overload;

function terminal_set(const Options: WideString): LongBool; overload;

// Refresh
procedure terminal_refresh();
  cdecl; external 'BearLibTerminal' name 'terminal_refresh';

// Clear
procedure terminal_clear();
  cdecl; external 'BearLibTerminal' name 'terminal_clear';

// ClearArea
procedure terminal_clear_area(Left, Top, Width, Height: Int32);
  cdecl; external 'BearLibTerminal' name 'terminal_clear_area';

// Crop
procedure terminal_crop(Left, Top, Width, Height: Int32);
  cdecl; external 'BearLibTerminal' name 'terminal_crop';

// Color
procedure terminal_color(Color: UInt32); overload;
  cdecl; external 'BearLibTerminal' name 'terminal_color';
  
procedure terminal_color(Color: AnsiString); overload;

procedure terminal_color(Color: WideString); overload;

// BkColor
procedure terminal_bkcolor(Color: UInt32); overload;
  cdecl; external 'BearLibTerminal' name 'terminal_bkcolor';

procedure terminal_bkcolor(Color: AnsiString); overload;

procedure terminal_bkcolor(Color: WideString); overload;

// Composition
procedure terminal_composition(Mode: Int32);
  cdecl; external 'BearLibTerminal' name 'terminal_composition';

// Layer
procedure terminal_layer(Mode: Int32);
  cdecl; external 'BearLibTerminal' name 'terminal_layer';

// Font
procedure terminal_font(const Name: AnsiString); overload;

procedure terminal_font(const Name: WideString); overload;

// Put
procedure terminal_put(X, Y, Code: Int32); overload;

procedure terminal_put(X, Y: Int32; Code: AnsiChar); overload;

procedure terminal_put(X, Y: Int32; Code: WideChar); overload;

// PutExt
procedure terminal_put_ext(X, Y, dX, dY: Int32; Code: Int32; Corners: PUInt32); overload;
  cdecl; external 'BearLibTerminal' name 'terminal_put_ext';

procedure terminal_put_ext(X, Y, dX, dY: Int32; Code: WideChar; Corners: PUInt32); overload;

procedure terminal_put_ext(X, Y, dX, dY: Int32; Code: Int32); overload;

procedure terminal_put_ext(X, Y, dX, dY: Int32; Code: WideChar); overload;

// Pick
function terminal_pick(X, Y, Index: Int32): Int32; overload;
  cdecl; external 'BearLibTerminal' name 'terminal_pick';

function terminal_pick(X, Y: Int32): Int32; overload;

// PickColor
function terminal_pick_color(X, Y, Index: Int32): UInt32; overload;
  cdecl; external 'BearLibTerminal' name 'terminal_pick_color';

function terminal_pick_color(X, Y: Int32): UInt32; overload;

// PickBkColor
function terminal_pick_bkcolor(X, Y: Int32): UInt32;
  cdecl; external 'BearLibTerminal' name 'terminal_pick_bkcolor';

// Print
function terminal_print(X, Y: Int32; const S: AnsiString): TSize; overload;

function terminal_print(X, Y: Int32; const S: WideString): TSize; overload;

function terminal_print(X, Y, Alignment: Int32; const S: AnsiString): TSize; overload;

function terminal_print(X, Y, Alignment: Int32; const S: WideString): TSize; overload;

function terminal_print(X, Y, W, H, Alignment: Int32; const S: AnsiString): TSize; overload;

function terminal_print(X, Y, W, H, Alignment: Int32; const S: WideString): TSize; overload;

// Measure
function terminal_measure(const S: AnsiString): TSize; overload;

function terminal_measure(const S: WideString): TSize; overload;

function terminal_measure(W, H: Int32; const S: AnsiString): TSize; overload;

function terminal_measure(W, H: Int32; const S: WideString): TSize; overload;

// HasInput
function terminal_has_input(): LongBool;
  cdecl; external 'BearLibTerminal' name 'terminal_has_input';

// State
function terminal_state(Code: Int32): Int32;
  cdecl; external 'BearLibTerminal' name 'terminal_state';

// Check
function terminal_check(Code: Int32): Boolean;

// Read
function terminal_read(): Int32;
  cdecl; external 'BearLibTerminal' name 'terminal_read';

// ReadStr
function terminal_read_str(X, Y: Int32; var S: AnsiString; MaxLength: Int32): Int32; overload;

function terminal_read_str(X, Y: Int32; var S: WideString; MaxLength: Int32): Int32; overload;
  
// Peek
function terminal_peek(): Int32;
  cdecl; external 'BearLibTerminal' name 'terminal_peek';

// Delay
procedure terminal_delay(Period: Int32);
  cdecl; external 'BearLibTerminal' name 'terminal_delay';

// Get
function terminal_get(const S: AnsiString): AnsiString; overload;

function terminal_get(const S, Default: AnsiString): AnsiString; overload;

function terminal_get(const S: WideString): WideString; overload;

function terminal_get(const S, Default: WideString): WideString; overload;

// ColorFromName
function color_from_name(const Name: AnsiString): UInt32; overload;

function color_from_name(const Name: WideString): UInt32; overload;

// ColorFromARGB
function color_from_argb(a, r, g, b: Int32): UInt32;

// ----------------------------------------------------------------------------
// Module implementation
// ----------------------------------------------------------------------------

implementation

function terminal_set_ansi(const Options: PAnsiChar): LongBool;
  cdecl; external 'BearLibTerminal' name 'terminal_set8';

function terminal_set(const Options: AnsiString): LongBool;
begin
    terminal_set := terminal_set_ansi(PAnsiChar(Options));
end;

function terminal_set_unicode(const Options: PWideChar): LongBool;
  cdecl; external 'BearLibTerminal' name 'terminal_set16';

function terminal_set(const Options: WideString): LongBool;
begin
    terminal_set := terminal_set_unicode(PWideChar(Options));
end;

procedure terminal_color(Color: AnsiString); overload;
begin
    terminal_color(color_from_name(Color));
end;

procedure terminal_color(Color: WideString); overload;
begin
    terminal_color(color_from_name(Color));
end;

procedure terminal_bkcolor(Color: AnsiString); overload;
begin
    terminal_bkcolor(color_from_name(Color));
end;

procedure terminal_bkcolor(Color: WideString); overload;
begin
    terminal_bkcolor(color_from_name(Color));
end;

procedure terminal_font_ansi(const Name: PAnsiChar);
  cdecl; external 'BearLibTerminal' name 'terminal_font8';

procedure terminal_font_unicode(const Name: PWideChar);
  cdecl; external 'BearLibTerminal' name 'terminal_font16';

procedure terminal_font(const Name: AnsiString); overload;
begin
	terminal_font_ansi(PAnsiChar(Name));
end;

procedure terminal_font(const Name: WideString); overload;
begin
	terminal_font_unicode(PWideChar(Name));
end;

procedure terminal_put_integer(X, Y: Int32; Code: Int32);
  cdecl; external 'BearLibTerminal' name 'terminal_put';

procedure terminal_put(X, Y, Code: Int32);
begin
    terminal_put_integer(X, Y, Code);
end;

procedure terminal_put(X, Y: Int32; Code: AnsiChar);
begin
    terminal_put_integer(X, Y, ord(Code));
end;

procedure terminal_put(X, Y: Int32; Code: WideChar);
begin
    terminal_put_integer(X, Y, ord(Code));
end;

procedure terminal_put_ext(X, Y, dX, dY: Int32; Code: WideChar; Corners: PUInt32); overload;
begin
    terminal_put_ext(X, Y, dX, dY, ord(Code), Corners);
end;

procedure terminal_put_ext(X, Y, dX, dY, Code: Int32); overload;
begin
    terminal_put_ext(X, Y, dX, dY, Code, PUInt32(0));
end;

procedure terminal_put_ext(X, Y, dX, dY: Int32; Code: WideChar); overload;
begin
    terminal_put_ext(X, Y, dX, dY, ord(Code), PUInt32(0));
end;

function terminal_pick(X, Y: Int32): Int32; overload;
begin
    terminal_pick := terminal_pick(X, Y, 0);
end;

function terminal_pick_color(X, Y: Int32): UInt32; overload;
begin;
    terminal_pick_color := terminal_pick_color(X, Y, 0);
end;

procedure terminal_print_ansi(X, Y, W, H, Alignment: Int32; const S: PAnsiChar; OutW, OutH: PInt32);
  cdecl; external 'BearLibTerminal' name 'terminal_print_ext8';
  
procedure terminal_print_unicode(X, Y, W, H, Alignment: Int32; const S: PWideChar; OutW, OutH: PInt32);
  cdecl; external 'BearLibTerminal' name 'terminal_print_ext16';

function terminal_print(X, Y: Int32; const S: AnsiString): TSize;
begin
    terminal_print := terminal_print(X, Y, 0, 0, TK_ALIGN_DEFAULT, S);
end;

function terminal_print(X, Y: Int32; const S: WideString): TSize;
begin
    terminal_print := terminal_print(X, Y, 0, 0, TK_ALIGN_DEFAULT, S);
end;

function terminal_print(X, Y, Alignment: Int32; const S: AnsiString): TSize;
begin
    terminal_print := terminal_print(X, Y, 0, 0, Alignment, S);
end;

function terminal_print(X, Y, Alignment: Int32; const S: WideString): TSize;
begin
    terminal_print := terminal_print(X, Y, 0, 0, Alignment, S);
end;

function terminal_print(X, Y, W, H, Alignment: Int32; const S: AnsiString): TSize;
var
    OutW, OutH: Int32;
begin
    terminal_print_ansi(X, Y, W, H, Alignment, PAnsiChar(S), @OutW, @OutH);
    terminal_print.cx := OutW;
    terminal_print.cy := OutH;
end;

function terminal_print(X, Y, W, H, Alignment: Int32; const S: WideString): TSize;
var
    OutW, OutH: Int32;
begin
    terminal_print_unicode(X, Y, W, H, Alignment, PWideChar(S), @OutW, @OutH);
    terminal_print.cx := OutW;
    terminal_print.cy := OutH;
end;

procedure terminal_measure_ansi(W, H: Int32; const S: PAnsiChar; OutW, OutH: PInt32);
  cdecl; external 'BearLibTerminal' name 'terminal_measure_ext8';

procedure terminal_measure_unicode(W, H: Int32; const S: PWideChar; OutW, OutH: PInt32);
  cdecl; external 'BearLibTerminal' name 'terminal_measure_ext16';

function terminal_measure(const S: AnsiString): TSize;
begin
    terminal_measure := terminal_measure(0, 0, S);
end;

function terminal_measure(const S: WideString): TSize;
begin
    terminal_measure := terminal_measure(0, 0, S);
end;

function terminal_measure(W, H: Int32; const S: AnsiString): TSize;
var
    OutW, OutH: Int32;
begin
    terminal_measure_ansi(W, H, PAnsiChar(S), @OutW, @OutH);
    terminal_measure.cx := OutW;
    terminal_measure.cy := OutH;
end;

function terminal_measure(W, H: Int32; const S: WideString): TSize;
var
    OutW, OutH: Int32;
begin
    terminal_measure_unicode(W, H, PWideChar(S), @OutW, @OutH);
    terminal_measure.cx := OutW;
    terminal_measure.cy := OutH;
end;

function terminal_check(Code: Int32): Boolean;
begin
    terminal_check := terminal_state(Code) > 0;
end;

function terminal_read_str_ansi(X, Y: Int32; S: PAnsiChar; MaxLength: Int32): Int32;
  cdecl; external 'BearLibTerminal' name 'terminal_read_str8';

function terminal_read_str(X, Y: Int32; var S: AnsiString; MaxLength: Int32): Int32;
var
    i: Integer;
begin
    SetLength(S, MaxLength*3+1); // Bigger size to accomodate possible UTF-8 string.
    terminal_read_str := terminal_read_str_ansi(X, Y, PAnsiChar(S), MaxLength);
    for i := 1 to MaxLength*3+1 do
        if S[i] = AnsiChar(0) then
            break;
    SetLength(S, i);
end;

function terminal_read_str_unicode(X, Y: Int32; S: PWideChar; MaxLength: Int32): Int32;
  cdecl; external 'BearLibTerminal' name 'terminal_read_str16';

function terminal_read_str(X, Y: Int32; var S: WideString; MaxLength: Int32): Int32;
var
    i: Integer;
begin
    SetLength(S, MaxLength);
    terminal_read_str := terminal_read_str_unicode(X, Y, PWideChar(S), MaxLength);
    for i := 1 to MaxLength do
        if S[i] = WideChar(0) then
            break;
    SetLength(S, i);
end;

function terminal_get_ansi(const S, Default: PAnsiChar): PAnsiChar;
  cdecl; external 'BearLibTerminal' name 'terminal_get8';

function terminal_get_unicode(const S, Default: PWideChar): PWideChar;
  cdecl; external 'BearLibTerminal' name 'terminal_get16';

function terminal_get(const S: AnsiString): AnsiString;
begin
    terminal_get := terminal_get_ansi(PAnsiChar(S), nil);
end;

function terminal_get(const S, Default: AnsiString): AnsiString;
begin
    terminal_get := terminal_get_ansi(PAnsiChar(S), PAnsiChar(Default));
end;

function terminal_get(const S: WideString): WideString;
begin
    terminal_get := terminal_get_unicode(PWideChar(S), nil);
end;

function terminal_get(const S, Default: WideString): WideString;
begin
    terminal_get := terminal_get_unicode(PWideChar(S), PWideChar(Default));
end;

function color_from_name_ansi(const Name: PAnsiChar): UInt32;
  cdecl; external 'BearLibTerminal' name 'color_from_name8';

function color_from_name(const Name: AnsiString): UInt32;
begin
    color_from_name := color_from_name_ansi(PAnsiChar(Name));
end;

function color_from_name_unicode(const Name: PWideChar): UInt32;
  cdecl; external 'BearLibTerminal' name 'color_from_name16';

function color_from_name(const Name: WideString): UInt32;
begin
    color_from_name := color_from_name_unicode(PWideChar(Name));
end;

function color_from_argb(a, r, g, b: Int32): UInt32;
begin
    if a > 255 then a := 255;
    if r > 255 then r := 255;
    if g > 255 then g := 255;
    if b > 255 then b := 255;
    color_from_argb := (a shl 24) or (r shl 16) or (g shl 8) or b;
end;

end.

File: Terminal/Source/BearLibTerminal.cpp
/*
* BearLibTerminal
* Copyright (C) 2013-2017 Cfyz
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
* of the Software, and to permit persons to whom the Software is furnished to do
* so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
* FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
* COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
* IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
* CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#define BEARLIBTERMINAL_BUILDING_LIBRARY
#include "BearLibTerminal.h"
#include "Config.hpp"
#include "Terminal.hpp"
#include "Palette.hpp"
#include "Config.hpp"
#include "Log.hpp"
#include <map>
#include <memory>
#include <string>
#include <string.h>
#include <iostream>
#include <thread>

using namespace BearLibTerminal;

namespace
{
	struct cached_setting_t
	{
		cached_setting_t();
		cached_setting_t(std::wstring s);
		cached_setting_t& operator=(std::wstring s);
		template<typename T> const T* get();

		std::string s8;
		std::u16string s16;
		std::u32string s32;
	};

	cached_setting_t::cached_setting_t()
	{ }

	cached_setting_t::cached_setting_t(std::wstring s)
	{
		this->operator=(s);
	}

	cached_setting_t& cached_setting_t::operator=(std::wstring s)
	{
		s8 = BearLibTerminal::UTF8Encoding().Convert(s);
		s16 = BearLibTerminal::UCS2Encoding().Convert(s);
		s32 = BearLibTerminal::UCS4Encoding().Convert(s);
		return *this;
	}

	template<> const char* cached_setting_t::get<char>()
	{
		return s8.c_str();
	}

	template<> const char16_t* cached_setting_t::get<char16_t>()
	{
		return s16.c_str();
	}

	template<> const char32_t* cached_setting_t::get<char32_t>()
	{
		return s32.c_str();
	}

	std::map<std::wstring, cached_setting_t> g_cached_settings;
}

int terminal_open()
{
	if (g_instance)
	{
		LOG(Error, "terminal_open: BearLibTerminal instance already initialized");
		return 0;
	}

	Config::Instance().Reload();
	Config::Instance().TryGet(L"ini.bearlibterminal.log.file", Log::Instance().filename);
	Config::Instance().TryGet(L"ini.bearlibterminal.log.level", Log::Instance().level);
	Config::Instance().TryGet(L"ini.bearlibterminal.log.mode", Log::Instance().mode);

	try
	{
		g_instance.reset(new BearLibTerminal::Terminal());
		return 1;
	}
	catch (std::exception& e)
	{
		LOG(Fatal, "terminal_open: " << e.what());
		return 0;
	}
}

void terminal_close()
{
	if (g_instance)
	{
		g_instance.reset();
		BearLibTerminal::Log::Instance().Reset();
	}
}

int terminal_set8(const int8_t* value)
{
	if (!g_instance || !value) return -1;
	auto& encoding = g_instance->GetEncoding();
	return g_instance->SetOptions(encoding.Convert((const char*)value));
}

int terminal_set16(const int16_t* value)
{
	if (!g_instance || !value) return -1;
	return g_instance->SetOptions(BearLibTerminal::UCS2Encoding().Convert((const char16_t*)value));
}

int terminal_set32(const int32_t* value)
{
	if (!g_instance || !value) return -1;
	return g_instance->SetOptions(BearLibTerminal::UCS4Encoding().Convert((const char32_t*)value));
}

void terminal_refresh()
{
	if (!g_instance) return;
	g_instance->Refresh();
}

void terminal_clear()
{
	if (!g_instance) return;
	g_instance->Clear();
}

void terminal_clear_area(int x, int y, int w, int h)
{
	if (!g_instance) return;
	g_instance->Clear(x, y, w, h);
}

void terminal_crop(int x, int y, int w, int h)
{
	if (!g_instance) return;
	g_instance->SetCrop(x, y, w, h);
}

void terminal_layer(int index)
{
	if (!g_instance) return;
	g_instance->SetLayer(index);
}

void terminal_color(color_t color)
{
	if (!g_instance) return;
	g_instance->SetForeColor(BearLibTerminal::Color(color));
}

void terminal_bkcolor(color_t color)
{
	if (!g_instance) return;
	g_instance->SetBackColor(BearLibTerminal::Color(color));
}

void terminal_composition(int mode)
{
	if (!g_instance) return;
	g_instance->SetComposition(mode);
}

void terminal_font8(const int8_t* name)
{
	if (g_instance)
		g_instance->SetFont(g_instance->GetEncoding().Convert((const char*)name));
}

void terminal_font16(const int16_t* name)
{
	if (g_instance)
		g_instance->SetFont(UCS2Encoding().Convert((const char16_t*)name));
}

void terminal_font32(const int32_t* name)
{
	if (g_instance)
		g_instance->SetFont(UCS4Encoding().Convert((const char32_t*)name));
}

void terminal_put(int x, int y, int code)
{
	if (!g_instance) return;
	g_instance->Put(x, y, code);
}

void terminal_put_ext(int x, int y, int dx, int dy, int code, color_t* corners)
{
	if (!g_instance) return;
	g_instance->PutExtended(x, y, dx, dy, code, (BearLibTerminal::Color*)corners);
}

int terminal_pick(int x, int y, int index)
{
	if (!g_instance) return 0;
	return g_instance->Pick(x, y, index);
}

color_t terminal_pick_color(int x, int y, int index)
{
	if (!g_instance) return 0;
	return g_instance->PickForeColor(x, y, index);
}

color_t terminal_pick_bkcolor(int x, int y)
{
	if (!g_instance) return 0;
	return g_instance->PickBackColor(x, y);
}

#define TERMINAL_PRINT_OR_MEASURE(x, y, a, cast, measure) \
	if (!g_instance || !s) { \
		if (out_w) *out_w = 0; \
		if (out_h) *out_h = 0; \
		return; \
	} \
	auto size = g_instance->Print(x, y, w, h, a, cast, false, measure); \
	if (out_w) *out_w = size.width; \
	if (out_h) *out_h = size.height;

void terminal_print_ext8(int x, int y, int w, int h, int align, const int8_t* s, int* out_w, int* out_h)
{
	TERMINAL_PRINT_OR_MEASURE(x, y, align, g_instance->GetEncoding().Convert((const char*)s), false)
}

void terminal_print_ext16(int x, int y, int w, int h, int align, const int16_t* s, int* out_w, int* out_h)
{
	TERMINAL_PRINT_OR_MEASURE(x, y, align, BearLibTerminal::UCS2Encoding().Convert((const char16_t*)s), false)
}

void terminal_print_ext32(int x, int y, int w, int h, int align, const int32_t* s, int* out_w, int* out_h)
{
	TERMINAL_PRINT_OR_MEASURE(x, y, align, BearLibTerminal::UCS4Encoding().Convert((const char32_t*)s), false)
}

void terminal_measure_ext8(int w, int h, const int8_t* s, int* out_w, int* out_h)
{
	TERMINAL_PRINT_OR_MEASURE(0, 0, TK_ALIGN_DEFAULT, g_instance->GetEncoding().Convert((const char*)s), true)
}

void terminal_measure_ext16(int w, int h, const int16_t* s, int* out_w, int* out_h)
{
	TERMINAL_PRINT_OR_MEASURE(0, 0, TK_ALIGN_DEFAULT, BearLibTerminal::UCS2Encoding().Convert((const char16_t*)s), true)
}

void terminal_measure_ext32(int w, int h, const int32_t* s, int* out_w, int* out_h)
{
	TERMINAL_PRINT_OR_MEASURE(0, 0, TK_ALIGN_DEFAULT, BearLibTerminal::UCS4Encoding().Convert((const char32_t*)s), true)
}

int terminal_has_input()
{
	if (!g_instance) return 1;
	return g_instance->HasInput();
}

int terminal_state(int code)
{
	if (!g_instance) return 0;
	return g_instance->GetState(code);
}

int terminal_read()
{
	if (!g_instance) return TK_CLOSE;
	return g_instance->Read();
}

template<typename char_t, typename enc_t> int read_str(int x, int y, char_t* buffer, int max, const enc_t& encoding)
{
	if (!g_instance) return -1;
	std::wstring wide_buffer = encoding.Convert((const char_t*)buffer);
	wide_buffer.reserve(max+1);
	int rc = g_instance->ReadString(x, y, &wide_buffer[0], max);
	if (rc >= 0)
	{
		std::basic_string<char_t> result = encoding.Convert(wide_buffer.c_str());
		memcpy(buffer, result.data(), sizeof(char_t)*(result.length()+1));
	}
	return rc;
}

int terminal_read_str8(int x, int y, int8_t* buffer, int max)
{
	if (!g_instance) return TK_INPUT_CANCELLED;
	return read_str(x, y, (char*)buffer, max, g_instance->GetEncoding());
}

int terminal_read_str16(int x, int y, int16_t* buffer, int max)
{
	if (!g_instance) return TK_INPUT_CANCELLED;
	return read_str(x, y, (char16_t*)buffer, max, BearLibTerminal::UCS2Encoding());
}

int terminal_read_str32(int x, int y, int32_t* buffer, int max)
{
	if (!g_instance) return TK_INPUT_CANCELLED;
	return read_str(x, y, (char32_t*)buffer, max, BearLibTerminal::UCS4Encoding());
}

int terminal_peek()
{
	if (!g_instance) return TK_CLOSE;
	return g_instance->Peek();
}

void terminal_delay(int period)
{
	if (g_instance)
		g_instance->Delay(period);
	else
		std::this_thread::sleep_for(std::chrono::milliseconds{period});
}

template<typename outer, typename inner> const outer* terminal_get(const outer* key, const outer* default_)
{
	typename BearLibTerminal::Encodings<inner>::type encoding;
	std::wstring wkey = encoding.Convert(std::basic_string<inner>((const inner*)key));
	std::wstring wout;
	if (!BearLibTerminal::Config::Instance().TryGet(wkey, wout))
	{
		wout = default_ == nullptr? std::wstring(L""): encoding.Convert(std::basic_string<inner>((const inner*)default_));
	}
	auto& cached_setting = g_cached_settings[wkey];
	cached_setting = wout;
	return (const outer*)cached_setting.get<inner>();
}

const int8_t* terminal_get8(const int8_t* key, const int8_t* default_)
{
	return terminal_get<int8_t, char>(key, default_);
}

const int16_t* terminal_get16(const int16_t* key, const int16_t* default_)
{
	return terminal_get<int16_t, char16_t>(key, default_);
}

const int32_t* terminal_get32(const int32_t* key, const int32_t* default_)
{
	return terminal_get<int32_t, char32_t>(key, default_);
}

color_t color_from_name8(const int8_t* name)
{
	if (!g_instance || !name) return -1;
	auto& encoding = g_instance->GetEncoding();
	return BearLibTerminal::Palette::Instance.Get(encoding.Convert((const char*)name));
}

color_t color_from_name16(const int16_t* name)
{
	if (!g_instance || !name) return -1;
	return BearLibTerminal::Palette::Instance.Get(BearLibTerminal::UCS2Encoding().Convert((const char16_t*)name));
}

color_t color_from_name32(const int32_t* name)
{
	if (!g_instance || !name) return -1;
	return BearLibTerminal::Palette::Instance.Get(BearLibTerminal::UCS4Encoding().Convert((const char32_t*)name));
}

int terminal_put_array(int x, int y, int w, int h, const uint8_t* data, int row_stride, int column_stride, const void* layout, int char_size)
{
	if (!g_instance)
	{
		return -1;
	}

	std::wstring s;
	switch (char_size)
	{
	case 1:
		s = g_instance->GetEncoding().Convert((const char*)layout);
		break;
	case 2:
		s = BearLibTerminal::UCS2Encoding().Convert((const char16_t*)layout);
		break;
	case 4:
		s = BearLibTerminal::UCS4Encoding().Convert((const char32_t*)layout);
		break;
	default:
		return -1;
	}

	return g_instance->PutArray(x, y, w, h, data, row_stride, column_stride, s);
}
